File: leetcode/top-k-frequent-elements.py
# 347. Top K Frequent Elements
# 🟠 Medium
#
# https://leetcode.com/problems/top-k-frequent-elements/
#
# Tags: Array - Hash Table - Divide and Conquer - Sorting
#  - Heap (Priority Queue) - Bucket Sort - Counting - Quickselect

import timeit
from collections import Counter
from typing import List


# Use the built-in Python Counter to count the elements and its
# `most_common(n)` method to get the k most frequent.
#
# Time complexity: O(n*log(n)) - Sorting the dictionary by value.
# Space complexity: O(n) - The size of the dictionary.
#
# Runtime 95 ms Beats 95.19%
# Memory Usage 18.4 Beats 99.66%
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        return [x[0] for x in Counter(nums).most_common(k)]


def test():
    executors = [Solution]
    tests = [
        [[1], 1, [1]],
        [[3, 0, 1, 0], 1, [0]],
        [[1, 1, 1, 2, 2, 3], 2, [1, 2]],
        [[1, 1, 1, 2, 2, 2, 2, 3], 3, [2, 1, 3]],
        [[1, 4, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 3, 4, 3], 3, [3, 2, 1]],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(1):
            for col, t in enumerate(tests):
                sol = executor()
                result = sol.topKFrequent(t[0], t[1])
                exp = t[2]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for"
                    + f" test {col} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/top-k-frequent-elements.rs
// 347. Top K Frequent Elements
// 🟠 Medium
//
// https://leetcode.com/problems/top-k-frequent-elements/
//
// Tags: Array - Hash Table - Divide and Conquer - Sorting - Heap (Priority Queue)
// - Bucket Sort - Counting - Quickselect

use std::collections::HashMap;

struct Solution;

/// Iterate over the elements to get their frequencies, then use a vector of
/// size m, where m is the maximum frequency of any element in the input, to
/// bucket sort the elements. Finally, iterate over the sorted elements and
/// pick the k most frequent to create the result vector.
///
/// Time complexity: O(n) - We iterate over all elements to create a hashmap of
/// their frequencies, then iterate over that hashmap and bucket sort them in
/// O(n), then iterate over the bucket sorted elements in O(k), where k <= n,
/// and create the result vector.
/// Space complexity: O(n) - The frequencies hashmap nad the buckets vector can
/// both grow to size n.
///
/// Runtime 2 ms Beats 85.57%
/// Memory 2.3 MB Beats 99.2%
impl Solution {
    pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let mut frequencies = HashMap::new();
        let mut highest_frequency = 0;
        for num in nums {
            *frequencies.entry(num).or_insert(0) += 1;
            highest_frequency = highest_frequency.max(*frequencies.get(&num).unwrap());
        }
        // Bucket sort the elements.
        let mut buckets = vec![vec![]; highest_frequency as usize + 1];
        for (key, val) in frequencies {
            buckets[val].push(key);
        }
        // Get the k most frequent.
        let k = k as usize;
        let mut res = Vec::with_capacity(k);
        let (mut i, mut j) = (buckets.len() - 1, 0);
        while res.len() < k {
            // Do we have an element at the current index?
            if j < buckets[i].len() {
                res.push(buckets[i][j]);
                j += 1;
            } else {
                i -= 1;
                j = 0;
            }
        }
        res
    }
}

// Tests.
fn main() {
    let tests = vec![
        (vec![1], 1, vec![1]),
        (vec![1, 1, 1, 2, 2, 3], 2, vec![1, 2]),
    ];
    for t in tests {
        assert_eq!(Solution::top_k_frequent(t.0, t.1), t.2);
    }
    println!("\x1b[92m» All tests passed!\x1b[0m")
}
